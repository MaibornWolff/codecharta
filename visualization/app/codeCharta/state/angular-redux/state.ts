import { Injectable } from "@angular/core"

import { Store as PlainStore } from "../store/store"

@Injectable({
	providedIn: "root"
})
export class State {
	getValue() {
		return PlainStore.store.getState()
	}
}
