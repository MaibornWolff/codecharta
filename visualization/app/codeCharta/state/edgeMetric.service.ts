import { MetricData, FileState, BlacklistItem, Edge, BlacklistType, CodeMapNode, EdgeMetricCount } from "../codeCharta.model"
import _ from "lodash"
import { FileStateServiceSubscriber, FileStateService } from "./fileState.service"
import { IRootScopeService } from "angular"
import { FileStateHelper } from "../util/fileStateHelper"
import { CodeMapHelper } from "../util/codeMapHelper"
import { BlacklistSubscriber } from "./settingsService/settings.service.events"
import { SettingsService } from "./settingsService/settings.service"
import { HierarchyNode } from "d3"

export interface EdgeMetricServiceSubscriber {
	onEdgeMetricDataUpdated(metricData: MetricData[])
}

export class EdgeMetricService implements FileStateServiceSubscriber, BlacklistSubscriber {
	private static EDGE_METRIC_DATA_UPDATED_EVENT = "edge-metric-data-updated"

	private edgeMetricData: MetricData[] = []
	private nodeEdgeMetricsMap: Map<string, Map<string, EdgeMetricCount>>

	constructor(private $rootScope: IRootScopeService, private fileStateService: FileStateService) {
		FileStateService.subscribe(this.$rootScope, this)
		SettingsService.subscribeToBlacklist(this.$rootScope, this)
	}

	public onBlacklistChanged(blacklist: BlacklistItem[]) {
		const fileStates: FileState[] = this.fileStateService.getFileStates()
		this.updateEdgeMetrics(fileStates, blacklist)
	}

	public onFileSelectionStatesChanged(fileStates: FileState[]) {
		this.updateEdgeMetrics(fileStates, [])
	}

	public onImportedFilesChanged(fileState: FileState[]) {}

	private updateEdgeMetrics(fileStates: FileState[], blacklist: BlacklistItem[]) {
		this.edgeMetricData = this.calculateMetrics(FileStateHelper.getVisibleFileStates(fileStates), blacklist)
		this.addNoneMetric()
		this.sortNodeEdgeMetricsMap()
		this.notifyEdgeMetricDataUpdated()
	}

	public getMetricNames(): string[] {
		if (!this.edgeMetricData) {
			return []
		}

		return this.edgeMetricData.map(x => x.name)
	}

	public getAmountOfAffectedBuildings(metricName: string): number {
		if (!this.nodeEdgeMetricsMap.has(metricName)) {
			return 0
		}
		return this.nodeEdgeMetricsMap.get(metricName).size
	}

	public getNodesWithHighestValue(metricName: string, numberOfNodes: number): string[] {
		if (!this.nodeEdgeMetricsMap.has(metricName)) {
			return []
		}

		let highestEdgeCountBuildings: string[] = []
		const edgeMetricMapKeyIterator = this.nodeEdgeMetricsMap.get(metricName).keys()
		for (let i = 0; i < numberOfNodes; i++) {
			highestEdgeCountBuildings.push(edgeMetricMapKeyIterator.next().value)
		}

		return highestEdgeCountBuildings
	}

	public getMetricValuesForNode(node: HierarchyNode<CodeMapNode>): Map<string, EdgeMetricCount> {
		const metricNames = this.getMetricNames().filter(it => !!this.nodeEdgeMetricsMap.get(it))
		let nodeEdgeMetrics = new Map()

		metricNames.forEach(metric => {
			nodeEdgeMetrics.set(metric, this.nodeEdgeMetricsMap.get(metric).get(node.data.path))
		})
		return nodeEdgeMetrics
	}

	public getMetricData(): MetricData[] {
		return this.edgeMetricData
	}

	private calculateMetrics(visibleFileStates: FileState[], blacklist: BlacklistItem[]): MetricData[] {
		if (visibleFileStates.length <= 0) {
			return []
		} else {
			const hashMap = this.calculateEdgeMetricData(visibleFileStates, blacklist)
			return this.getMetricDataFromMap(hashMap)
		}
	}

	private calculateEdgeMetricData(fileStates: FileState[], blacklist: BlacklistItem[]): Map<string, Map<string, EdgeMetricCount>> {
		this.nodeEdgeMetricsMap = new Map()
		fileStates.forEach(fileState => {
			fileState.file.settings.fileSettings.edges.forEach(edge => {
				if (this.bothNodesAssociatedAreVisible(edge, blacklist, fileStates)) {
					this.addEdgeToCalculationMap(edge)
				}
			})
		})
		return this.nodeEdgeMetricsMap
	}

	private bothNodesAssociatedAreVisible(edge: Edge, blacklist: BlacklistItem[], fileStates: FileState[]): boolean {
		const fromNode = this.getMapNodeFromPath(edge.fromNodeName, fileStates)
		const toNode = this.getMapNodeFromPath(edge.toNodeName, fileStates)
		return fromNode && toNode && this.isNotBlacklisted(fromNode, blacklist) && this.isNotBlacklisted(toNode, blacklist)
	}

	private getMapNodeFromPath(path: string, fileStates: FileState[]): CodeMapNode {
		let mapNode = undefined
		fileStates.forEach(fileState => {
			const nodeFound = CodeMapHelper.getCodeMapNodeFromPath(path, "File", fileState.file.map)
			if (nodeFound) {
				mapNode = nodeFound
			}
		})
		return mapNode
	}

	private isNotBlacklisted(node: CodeMapNode, blacklist: BlacklistItem[]): boolean {
		return !CodeMapHelper.isBlacklisted(node, blacklist, BlacklistType.exclude)
	}

	private addEdgeToCalculationMap(edge: Edge) {
		for (let edgeMetric of Object.keys(edge.attributes)) {
			const edgeMetricEntry = this.getEntryForMetric(edgeMetric)
			this.addEdgeToNodes(edgeMetricEntry, edge.fromNodeName, edge.toNodeName)
		}
	}

	private getEntryForMetric(edgeMetricName: string): Map<string, EdgeMetricCount> {
		if (!this.nodeEdgeMetricsMap.has(edgeMetricName)) {
			this.nodeEdgeMetricsMap.set(edgeMetricName, new Map())
		}
		return this.nodeEdgeMetricsMap.get(edgeMetricName)
	}

	private addEdgeToNodes(edgeMetricEntry: Map<string, EdgeMetricCount>, fromNode: string, toNode: string) {
		this.createEntryIfNecessary(edgeMetricEntry, fromNode)
		this.createEntryIfNecessary(edgeMetricEntry, toNode)
		edgeMetricEntry.get(fromNode).outgoing += 1
		edgeMetricEntry.get(toNode).incoming += 1
	}

	private createEntryIfNecessary(edgeMetricEntry: Map<string, EdgeMetricCount>, nodeName: string) {
		if (!edgeMetricEntry.has(nodeName)) {
			edgeMetricEntry.set(nodeName, { incoming: 0, outgoing: 0 })
		}
	}

	private getMetricDataFromMap(hashMap: Map<string, Map<string, EdgeMetricCount>>): MetricData[] {
		let metricData: MetricData[] = []

		hashMap.forEach((occurences: any, edgeMetric: any) => {
			let maximumMetricValue = 0
			occurences.forEach((value: EdgeMetricCount, _) => {
				const combinedValue = value.incoming + value.outgoing
				if (combinedValue > maximumMetricValue) {
					maximumMetricValue = combinedValue
				}
			})
			metricData.push({ name: edgeMetric, maxValue: maximumMetricValue, availableInVisibleMaps: true })
		})

		return metricData
	}

	private sortNodeEdgeMetricsMap() {
		let sortedEdgeMetricMap = new Map()
		this.nodeEdgeMetricsMap.forEach((value, key) => {
			const sortedMapForMetric = new Map(
				[...value.entries()].sort((a, b) => b[1].incoming + b[1].outgoing - (a[1].incoming + a[1].outgoing))
			)
			sortedEdgeMetricMap.set(key, sortedMapForMetric)
		})
		this.nodeEdgeMetricsMap = sortedEdgeMetricMap
	}

	private addNoneMetric() {
		this.edgeMetricData.push({ name: "None", maxValue: 0, availableInVisibleMaps: false })
	}

	private notifyEdgeMetricDataUpdated() {
		this.$rootScope.$broadcast(EdgeMetricService.EDGE_METRIC_DATA_UPDATED_EVENT, this.edgeMetricData)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: EdgeMetricServiceSubscriber) {
		$rootScope.$on(EdgeMetricService.EDGE_METRIC_DATA_UPDATED_EVENT, (event, data) => {
			subscriber.onEdgeMetricDataUpdated(data)
		})
	}
}
