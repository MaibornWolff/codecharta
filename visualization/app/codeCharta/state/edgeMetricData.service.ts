import { MetricData, BlacklistItem, Edge, BlacklistType, CodeMapNode, EdgeMetricCount, AttributeTypeValue } from "../codeCharta.model"
import { IRootScopeService } from "angular"
import { CodeMapHelper } from "../util/codeMapHelper"
import { HierarchyNode } from "d3"
import { BlacklistService, BlacklistSubscriber } from "./store/fileSettings/blacklist/blacklist.service"
import { FilesService, FilesSelectionSubscriber } from "./store/files/files.service"
import { Files } from "../model/files"
import { StoreService } from "./store.service"

export interface EdgeMetricDataServiceSubscriber {
	onEdgeMetricDataUpdated(metricData: MetricData[])
}

export class EdgeMetricDataService implements FilesSelectionSubscriber, BlacklistSubscriber {
	private static EDGE_METRIC_DATA_UPDATED_EVENT = "edge-metric-data-updated"

	private edgeMetricData: MetricData[] = []
	private nodeEdgeMetricsMap: Map<string, Map<string, EdgeMetricCount>>

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		FilesService.subscribe(this.$rootScope, this)
		BlacklistService.subscribe(this.$rootScope, this)
	}

	public onBlacklistChanged(blacklist: BlacklistItem[]) {
		this.updateEdgeMetrics()
	}

	public onFilesSelectionChanged(files: Files) {
		this.updateEdgeMetrics()
	}

	private updateEdgeMetrics() {
		this.edgeMetricData = this.calculateMetrics()
		this.addNoneMetric()
		this.sortNodeEdgeMetricsMap()
		this.notifyEdgeMetricDataUpdated()
	}

	public getMetricNames(): string[] {
		if (!this.edgeMetricData) {
			return []
		}

		return this.edgeMetricData.map(x => x.name)
	}

	public getAmountOfAffectedBuildings(metricName: string): number {
		if (!this.nodeEdgeMetricsMap || !this.nodeEdgeMetricsMap.has(metricName)) {
			return 0
		}
		return this.nodeEdgeMetricsMap.get(metricName).size
	}

	public getNodesWithHighestValue(metricName: string, numberOfNodes: number): string[] {
		if (!this.nodeEdgeMetricsMap || !this.nodeEdgeMetricsMap.has(metricName)) {
			return []
		}

		let highestEdgeCountBuildings: string[] = []
		const edgeMetricMapKeyIterator = this.nodeEdgeMetricsMap.get(metricName).keys()
		for (let i = 0; i < numberOfNodes; i++) {
			highestEdgeCountBuildings.push(edgeMetricMapKeyIterator.next().value)
		}

		return highestEdgeCountBuildings
	}

	public getMetricValuesForNode(node: HierarchyNode<CodeMapNode>): Map<string, EdgeMetricCount> {
		const metricNames = this.getMetricNames().filter(it => !!this.nodeEdgeMetricsMap.get(it))
		let nodeEdgeMetrics = new Map()

		metricNames.forEach(metric => {
			nodeEdgeMetrics.set(metric, this.nodeEdgeMetricsMap.get(metric).get(node.data.path))
		})
		return nodeEdgeMetrics
	}

	public getMetricData(): MetricData[] {
		return this.edgeMetricData
	}

	public getAttributeTypeByMetric(metricName: string): AttributeTypeValue {
		return this.storeService.getState().fileSettings.attributeTypes.edges[metricName]
	}

	private calculateMetrics(): MetricData[] {
		if (!this.storeService.getState().files.fileStatesAvailable()) {
			return []
		} else {
			const hashMap = this.calculateEdgeMetricData()
			return this.getMetricDataFromMap(hashMap)
		}
	}

	private calculateEdgeMetricData(): Map<string, Map<string, EdgeMetricCount>> {
		this.nodeEdgeMetricsMap = new Map()
		const pathsPerFileState = this.storeService
			.getState()
			.files.getVisibleFileStates()
			.map(fileState => CodeMapHelper.getAllPaths(fileState.file.map))
		const allFilePaths: string[] = [].concat(...pathsPerFileState)
		this.storeService
			.getState()
			.files.getVisibleFileStates()
			.forEach(fileState => {
				fileState.file.settings.fileSettings.edges.forEach(edge => {
					if (this.bothNodesAssociatedAreVisible(edge, allFilePaths)) {
						this.addEdgeToCalculationMap(edge)
					}
				})
			})
		return this.nodeEdgeMetricsMap
	}

	private bothNodesAssociatedAreVisible(edge: Edge, filePaths: string[]): boolean {
		const fromPath = filePaths.find(x => x === edge.fromNodeName)
		const toPath = filePaths.find(x => x === edge.toNodeName)
		return fromPath && toPath && this.isNotBlacklisted(fromPath) && this.isNotBlacklisted(toPath)
	}

	private isNotBlacklisted(path: string): boolean {
		const blacklist = this.storeService.getState().fileSettings.blacklist
		return !CodeMapHelper.isPathBlacklisted(path, blacklist, BlacklistType.exclude)
	}

	private addEdgeToCalculationMap(edge: Edge) {
		for (let edgeMetric of Object.keys(edge.attributes)) {
			const edgeMetricEntry = this.getEntryForMetric(edgeMetric)
			this.addEdgeToNodes(edgeMetricEntry, edge.fromNodeName, edge.toNodeName)
		}
	}

	private getEntryForMetric(edgeMetricName: string): Map<string, EdgeMetricCount> {
		if (!this.nodeEdgeMetricsMap.has(edgeMetricName)) {
			this.nodeEdgeMetricsMap.set(edgeMetricName, new Map())
		}
		return this.nodeEdgeMetricsMap.get(edgeMetricName)
	}

	private addEdgeToNodes(edgeMetricEntry: Map<string, EdgeMetricCount>, fromNode: string, toNode: string) {
		this.createEntryIfNecessary(edgeMetricEntry, fromNode)
		this.createEntryIfNecessary(edgeMetricEntry, toNode)
		edgeMetricEntry.get(fromNode).outgoing += 1
		edgeMetricEntry.get(toNode).incoming += 1
	}

	private createEntryIfNecessary(edgeMetricEntry: Map<string, EdgeMetricCount>, nodeName: string) {
		if (!edgeMetricEntry.has(nodeName)) {
			edgeMetricEntry.set(nodeName, { incoming: 0, outgoing: 0 })
		}
	}

	private getMetricDataFromMap(hashMap: Map<string, Map<string, EdgeMetricCount>>): MetricData[] {
		let metricData: MetricData[] = []

		hashMap.forEach((occurences: any, edgeMetric: any) => {
			let maximumMetricValue = 0
			occurences.forEach((value: EdgeMetricCount, _) => {
				const combinedValue = value.incoming + value.outgoing
				if (combinedValue > maximumMetricValue) {
					maximumMetricValue = combinedValue
				}
			})
			metricData.push({ name: edgeMetric, maxValue: maximumMetricValue })
		})

		return metricData
	}

	private sortNodeEdgeMetricsMap() {
		let sortedEdgeMetricMap = new Map()
		if (this.nodeEdgeMetricsMap) {
			this.nodeEdgeMetricsMap.forEach((value, key) => {
				const sortedMapForMetric = new Map(
					[...value.entries()].sort((a, b) => b[1].incoming + b[1].outgoing - (a[1].incoming + a[1].outgoing))
				)
				sortedEdgeMetricMap.set(key, sortedMapForMetric)
			})
		}
		this.nodeEdgeMetricsMap = sortedEdgeMetricMap
	}

	private addNoneMetric() {
		this.edgeMetricData.push({ name: "None", maxValue: 0 })
	}

	private notifyEdgeMetricDataUpdated() {
		this.$rootScope.$broadcast(EdgeMetricDataService.EDGE_METRIC_DATA_UPDATED_EVENT, this.edgeMetricData)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: EdgeMetricDataServiceSubscriber) {
		$rootScope.$on(EdgeMetricDataService.EDGE_METRIC_DATA_UPDATED_EVENT, (event, data) => {
			subscriber.onEdgeMetricDataUpdated(data)
		})
	}
}
