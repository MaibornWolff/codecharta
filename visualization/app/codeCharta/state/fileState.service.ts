import { CCFile, FileSelectionState, FileState } from "../codeCharta.model"
import { IRootScopeService } from "angular"
import { LoadingGifService } from "../ui/loadingGif/loadingGif.service"

export interface FileStateServiceSubscriber {
	onFileSelectionStatesChanged(fileStates: FileState[])

	onImportedFilesChanged(fileStates: FileState[])
}

export class FileStateService {
	private static FILE_STATE_CHANGED_EVENT = "file-selection-states-changed"
	private static IMPORTED_FILES_CHANGED_EVENT = "imported-files-changed"

	private fileStates: Array<FileState> = []

	/* @ngInject */
	constructor(private $rootScope: IRootScopeService, private loadingGifService: LoadingGifService) {}

	public resetMaps() {
		this.fileStates = []
	}

	public addFile(file: CCFile) {
		this.fileStates.push({ file: file, selectedAs: FileSelectionState.None })
		this.notifyFileImport()
	}

	public getCCFiles(): CCFile[] {
		return this.fileStates.map(x => x.file)
	}

	public getFileStates(): FileState[] {
		return this.fileStates
	}

	public setSingle(file: CCFile) {
		this.resetSelectionStates()
		const matchedFile = this.fileStates.find(x => x.file === file)
		if (matchedFile) {
			matchedFile.selectedAs = FileSelectionState.Single
		}
		this.notifySelectionChange()
	}

	public setDelta(reference: CCFile, comparison: CCFile) {
		this.resetSelectionStates()
		const matchedReferenceFile = this.fileStates.find(x => x.file === reference)
		const matchedComparisonFile = this.fileStates.find(x => x.file === comparison)

		if (matchedReferenceFile) {
			matchedReferenceFile.selectedAs = FileSelectionState.Reference
		}

		if (matchedComparisonFile) {
			matchedComparisonFile.selectedAs = FileSelectionState.Comparison
		}
		this.notifySelectionChange()
	}

	public setMultiple(multipleFiles: CCFile[]) {
		this.resetSelectionStates()
		this.fileStates.filter(x => multipleFiles.indexOf(x.file) !== -1).forEach(x => (x.selectedAs = FileSelectionState.Partial))
		this.notifySelectionChange()
	}

	private resetSelectionStates() {
		this.fileStates.forEach(file => (file.selectedAs = FileSelectionState.None))
	}

	private notifySelectionChange() {
		this.loadingGifService.updateLoadingMapFlag(true)
		this.$rootScope.$broadcast(FileStateService.FILE_STATE_CHANGED_EVENT, this.fileStates)
	}

	private notifyFileImport() {
		this.$rootScope.$broadcast(FileStateService.IMPORTED_FILES_CHANGED_EVENT, this.fileStates)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: FileStateServiceSubscriber) {
		$rootScope.$on(FileStateService.FILE_STATE_CHANGED_EVENT, (event, data) => {
			subscriber.onFileSelectionStatesChanged(data)
		})
		$rootScope.$on(FileStateService.IMPORTED_FILES_CHANGED_EVENT, (event, data) => {
			subscriber.onImportedFilesChanged(data)
		})
	}
}
