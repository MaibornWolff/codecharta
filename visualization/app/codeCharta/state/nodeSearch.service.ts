import { CodeMapNode } from "../codeCharta.model"
import { IRootScopeService } from "angular"
import { CodeMapHelper } from "../util/codeMapHelper"
import { CodeMapPreRenderService } from "../ui/codeMap/codeMap.preRender.service"
import { StoreService } from "./store.service"
import { setSearchedNodePaths } from "./store/dynamicSettings/searchedNodePaths/searchedNodePaths.actions"
import { SearchPatternService, SearchPatternSubscriber } from "./store/dynamicSettings/searchPattern/searchPattern.service"

export interface NodeSearchSubscriber {
	onNodeSearchComplete(searchedNodes: CodeMapNode[])
}

export class NodeSearchService implements SearchPatternSubscriber {
	private static NODE_SEARCH_COMPLETE_EVENT = "node-search-complete"

	private searchedNodes: CodeMapNode[] = []

	/* @ngInject */
	constructor(
		private $rootScope: IRootScopeService,
		private storeService: StoreService,
		private codeMapPreRenderService: CodeMapPreRenderService
	) {
		SearchPatternService.subscribe(this.$rootScope, this)
	}

	onSearchPatternChanged(searchPattern: string) {
		this.searchedNodes = CodeMapHelper.getNodesByGitignorePath(this.codeMapPreRenderService.getRenderMap(), searchPattern)
		this.notifyNodeSearchComplete()
		this.applySettingsSearchedNodePaths()
	}

	private applySettingsSearchedNodePaths() {
		const newSearchedNodePaths = this.searchedNodes.map(x => x.path)
		this.storeService.dispatch(setSearchedNodePaths(new Set(newSearchedNodePaths)))
	}

	private notifyNodeSearchComplete() {
		this.$rootScope.$broadcast(NodeSearchService.NODE_SEARCH_COMPLETE_EVENT, this.searchedNodes)
	}

	static subscribe($rootScope: IRootScopeService, subscriber: NodeSearchSubscriber) {
		$rootScope.$on(NodeSearchService.NODE_SEARCH_COMPLETE_EVENT, (_event_, data) => {
			subscriber.onNodeSearchComplete(data)
		})
	}
}
