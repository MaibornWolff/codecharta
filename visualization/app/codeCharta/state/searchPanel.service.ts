import { IRootScopeService } from "angular"
import { SearchPanelMode } from "../codeCharta.model"

export interface SearchPanelServiceSubscriber {
	onSearchPanelModeChanged(searchPanelMode: SearchPanelMode)
}

export class SearchPanelService {
	public static readonly SEARCH_PANEL_MODE_EVENT = "search-panel-mode-changed"

	private searchPanelMode: SearchPanelMode = SearchPanelMode.minimized

	/* @ngInject */
	constructor(private $rootScope: IRootScopeService) {}

	public updateSearchPanelMode(searchPanelMode: SearchPanelMode) {
		this.searchPanelMode = searchPanelMode
		this.notifySearchPanelModeChanged()
	}

	private notifySearchPanelModeChanged() {
		this.$rootScope.$broadcast(SearchPanelService.SEARCH_PANEL_MODE_EVENT, this.searchPanelMode)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: SearchPanelServiceSubscriber) {
		$rootScope.$on(SearchPanelService.SEARCH_PANEL_MODE_EVENT, (event, data) => {
			subscriber.onSearchPanelModeChanged(data)
		})
	}
}
