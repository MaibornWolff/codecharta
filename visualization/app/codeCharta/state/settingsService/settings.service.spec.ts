import "../state.module"
import { SettingsService } from "./settings.service"
import { IRootScopeService, ITimeoutService } from "angular"
import { LoadingGifService } from "../../ui/loadingGif/loadingGif.service"
import { AttributeTypeValue, FileSelectionState, FileState, RecursivePartial, Settings } from "../../codeCharta.model"
import { getService, instantiateModule } from "../../../../mocks/ng.mockhelper"
import { DEFAULT_SETTINGS, TEST_DELTA_MAP_A, TEST_DELTA_MAP_B } from "../../util/dataMocks"
import { FileStateService } from "../fileState.service"
import { FileStateHelper } from "../../util/fileStateHelper"
import { SettingsMerger } from "../../util/settingsMerger"

describe("settingService", () => {
	let settingsService: SettingsService
	let $rootScope: IRootScopeService
	let $timeout: ITimeoutService
	let loadingGifService: LoadingGifService

	let settings: Settings
	let fileStates: FileState[]
	const SOME_EXTRA_TIME = 400

	beforeEach(() => {
		restartSystem()
		rebuildService()
		withMockedEventMethods()
		withMockedLoadingGifService()
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.state")

		$rootScope = getService<IRootScopeService>("$rootScope")
		$timeout = getService<ITimeoutService>("$timeout")
		loadingGifService = getService<LoadingGifService>("loadingGifService")

		settings = JSON.parse(JSON.stringify(DEFAULT_SETTINGS))
		fileStates = [
			{ file: JSON.parse(JSON.stringify(TEST_DELTA_MAP_A)), selectedAs: FileSelectionState.Comparison },
			{ file: JSON.parse(JSON.stringify(TEST_DELTA_MAP_B)), selectedAs: FileSelectionState.Reference }
		]
	}

	function rebuildService() {
		settingsService = new SettingsService($rootScope, $timeout, loadingGifService)
	}

	function withMockedEventMethods() {
		$rootScope.$on = settingsService["$rootScope"].$on = jest.fn()
		$rootScope.$broadcast = settingsService["$rootScope"].$on = jest.fn()
	}

	function withMockedLoadingGifService() {
		loadingGifService = settingsService["loadingGifService"] = jest.fn().mockReturnValue({
			updateLoadingMapFlag: jest.fn()
		})()
	}

	describe("constructor", () => {
		it("should set settings to default settings", () => {
			rebuildService()

			expect(settingsService.getSettings()).toEqual(settings)
		})

		it("should subscribe to FileStateService", () => {
			FileStateService.subscribe = jest.fn()

			rebuildService()

			expect(FileStateService.subscribe).toHaveBeenCalledWith($rootScope, settingsService)
		})
	})

	describe("FileStateServiceSubscriber Methods", () => {
		beforeEach(() => {
			settingsService.updateSettings = jest.fn()
			FileStateHelper.isPartialState = jest.fn().mockReturnValue(false)
			FileStateHelper.getVisibleFileStates = jest.fn().mockReturnValue(fileStates)
			SettingsMerger.getMergedFileSettings = jest.fn().mockReturnValue(DEFAULT_SETTINGS)
		})

		describe("onFileSelectionStateChanged", () => {
			it("should call updateSettings with newFileSettings", () => {
				settingsService.onFileSelectionStatesChanged(fileStates)

				expect(settingsService.updateSettings).toHaveBeenCalledWith({ fileSettings: DEFAULT_SETTINGS })
			})

			it("should call isPartialState with fileStates", () => {
				settingsService.onFileSelectionStatesChanged(fileStates)

				expect(FileStateHelper.isPartialState).toHaveBeenCalledWith(fileStates)
			})

			it("should call getVisibleFileStates with fileStates", () => {
				settingsService.onFileSelectionStatesChanged(fileStates)

				expect(FileStateHelper.getVisibleFileStates).toHaveBeenCalledWith(fileStates)
			})

			it("should call getMergedFileStates with visibleFiles and withUpdatedPath", () => {
				settingsService.onFileSelectionStatesChanged(fileStates)
				const visibleFiles = [fileStates[0].file, fileStates[1].file]

				expect(SettingsMerger.getMergedFileSettings).toHaveBeenCalledWith(visibleFiles, false)
			})
		})
	})

	describe("updateSettings", () => {
		it("should set settings correctly", () => {
			const expected = {
				...DEFAULT_SETTINGS,
				appSettings: { ...DEFAULT_SETTINGS.appSettings, invertHeight: true }
			}

			settingsService.updateSettings({ appSettings: { invertHeight: true } })

			expect(settingsService.getSettings()).toEqual(expected)
		})

		it("should call updateLoadingMapFlag", () => {
			settingsService.updateSettings({ appSettings: { invertHeight: true } })

			expect(loadingGifService.updateLoadingMapFlag).toHaveBeenCalledWith(true)
		})

		it("should set attributeTypes", () => {
			const nodeAttributeTypes = [
				{ rloc: AttributeTypeValue.absolute },
				{ mcc: AttributeTypeValue.absolute },
				{ coverage: AttributeTypeValue.relative }
			]

			settingsService.updateSettings({
				fileSettings: {
					attributeTypes: {
						nodes: nodeAttributeTypes
					}
				}
			})

			expect(settingsService.getSettings().fileSettings.attributeTypes.nodes).toEqual(nodeAttributeTypes)
		})

		it("should merge two objects with different root properties", () => {
			settingsService["update"] = { dynamicSettings: { areaMetric: "rloc" } }
			const update = { appSettings: { margin: 2 } } as RecursivePartial<Settings>
			const expected = {
				dynamicSettings: { areaMetric: "rloc" },
				appSettings: { margin: 2 }
			}

			settingsService.updateSettings(update)

			expect(settingsService["update"]).toEqual(expected)
		})

		it("should merge two objects with different non-root properties", () => {
			settingsService["update"] = { dynamicSettings: { areaMetric: "rloc" } }
			const update: RecursivePartial<Settings> = { dynamicSettings: { colorMetric: "mcc" } }
			const expected = {
				dynamicSettings: { areaMetric: "rloc", colorMetric: "mcc" }
			}

			settingsService.updateSettings(update)

			expect(settingsService["update"]).toEqual(expected)
		})

		it("should merge two objects and override one property", () => {
			settingsService["update"] = { dynamicSettings: { areaMetric: "rloc" } }
			const update: RecursivePartial<Settings> = { dynamicSettings: { areaMetric: "mcc" } }
			const expected = {
				dynamicSettings: { areaMetric: "mcc" }
			}

			settingsService.updateSettings(update)

			expect(settingsService["update"]).toEqual(expected)
		})

		it("should merge two objects and override arrays", () => {
			settingsService["update"] = { fileSettings: { blacklist: [] } }
			const update: RecursivePartial<Settings> = { fileSettings: { blacklist: ["entry"] } }
			const expected = {
				fileSettings: { blacklist: ["entry"] }
			}

			settingsService.updateSettings(update)

			expect(settingsService["update"]).toEqual(expected)
		})

		it("should merge two objects and override arrays", () => {
			settingsService["update"] = { fileSettings: { blacklist: ["entry"] } }
			const update: RecursivePartial<Settings> = { fileSettings: { blacklist: [] } }
			const expected = {
				fileSettings: { blacklist: [] }
			}

			settingsService.updateSettings(update)

			expect(settingsService["update"]).toEqual(expected)
		})

		it("should reset update after 400ms", done => {
			settingsService["update"] = { fileSettings: { edges: ["entry"] } }
			const update: RecursivePartial<Settings> = { fileSettings: { edges: [] } }

			settingsService.updateSettings(update)

			setTimeout(() => {
				expect(settingsService["update"]).toEqual({})
				done()
			}, SettingsService["DEBOUNCE_TIME"] + SOME_EXTRA_TIME)
		})
	})

	describe("getDefaultSettings", () => {
		it("should match snapshot", () => {
			expect(settingsService.getDefaultSettings()).toMatchSnapshot()
		})
	})

	describe("subscribe", () => {
		it("should setup one event listener", () => {
			SettingsService.subscribe($rootScope, undefined)

			expect($rootScope.$on).toHaveBeenCalledTimes(1)
		})
	})
})
