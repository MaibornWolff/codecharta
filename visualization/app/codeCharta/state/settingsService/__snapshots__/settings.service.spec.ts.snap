// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settingService getDefaultSettings should match snapshot 1`] = `
Object {
  "appSettings": Object {
    "amountOfEdgePreviews": 1,
    "amountOfTopLabels": 1,
    "camera": Vector3 {
      "x": 0,
      "y": 300,
      "z": 1000,
    },
    "dynamicMargin": true,
    "edgeHeight": 4,
    "hideFlatBuildings": false,
    "invertColorRange": false,
    "invertDeltaColors": false,
    "invertHeight": false,
    "isPresentationMode": false,
    "isWhiteBackground": false,
    "mapColors": Object {
      "angularGreen": "#00BFA5",
      "base": "#666666",
      "defaultC": "#89ACB4",
      "flat": "#AAAAAA",
      "incomingEdge": "#00ffff",
      "lightGrey": "#DDDDDD",
      "markingColors": Array [
        "#FF1D8E",
        "#1d8eff",
        "#1DFFFF",
        "#8eff1d",
        "#8e1dff",
        "#FFFF1D",
      ],
      "negative": "#820E0E",
      "negativeDelta": "#ff0E0E",
      "neutral": "#ddcc00",
      "outgoingEdge": "#ff00ff",
      "positive": "#69AE40",
      "positiveDelta": "#69FF40",
      "selected": "#EB8319",
    },
    "resetCameraIfNewFileIsLoaded": true,
    "scaling": Vector3 {
      "x": 1,
      "y": 1,
      "z": 1,
    },
    "showOnlyBuildingsWithEdges": false,
    "whiteColorBuildings": false,
  },
  "dynamicSettings": Object {
    "areaMetric": null,
    "colorMetric": null,
    "colorRange": Object {
      "from": null,
      "to": null,
    },
    "distributionMetric": null,
    "edgeMetric": null,
    "focusedNodePath": "",
    "heightMetric": null,
    "margin": null,
    "searchPattern": "",
    "searchedNodePaths": Array [],
  },
  "fileSettings": Object {
    "attributeTypes": Object {
      "edges": Array [],
      "nodes": Array [],
    },
    "blacklist": Array [],
    "edges": Array [],
    "markedPackages": Array [],
  },
  "treeMapSettings": Object {
    "mapSize": 250,
  },
}
`;
